### ISSUE TITLE
Unreliability of ZRTP

### ISSUE THREAD

**Issue by jrsnen** [ultravideo/uvgRTP#15]
## Bug Description
The ZRTP fails around 40% in localhost, which seems concerning. I think the ZRTP protocol should be robust enough not to fail consistently in certain situations. Timing seems to affect this bug (how fast each side is started) but I'm not certain. The ZRTP should also be tested repeatedly over other types of connections. I have indications that ZRTP can also fail when used over the internet, but it is possible that this is a different issue.

![image](https://user-images.githubusercontent.com/6498638/156166946-421063ff-c41e-4bd6-af6b-8d811a4ef9f5.png)

## Platforms
I have replicated this both on Windows and Linux.

## How to replicate
This can be replicated by running the ZRTP multistream example or by running the automated tests. Both fail sometimes.

## Solution

I'm not familiar enough with the protocol to say exactly how to fix this issue, other than trying to make the protocol implementation more robust somehow. One thing to look at is the mutexes in media_stream and whether they can lock one of the threads out while other is performing the ZRTP and failing.

I've previously traced it to uvgrtp::zrtp::begin_session() timeout at this line: https://github.com/ultravideo/uvgRTP/blob/b0b9ea6dfb75b99dbc481f2f89d7979b58253dd5/src/zrtp.cc#L405 

I tried increasing timeout and number of hello attempts, but this did not help. It clearly waited longer, but still did not manage to complete ZRTP.

Recent improvements to reception architecture seem to have mitigated the issue somewhat from what it was before, but I can still replicate the issue. 

## Affected files

Basically anything with zrtp in its name/path and media_stream.cc

**Comment by altonen** [comment]
Hard to say what the issue is, especially since it happens only sometimes. Maybe Hello/HelloACK is not received at all but I don't know why that would happen either.

**Comment by polusto** [comment]
I have also replicated this issue on Windows while improving the ZRTP multistream example.

```
[RTPLIB][DEBUG][::alloc_zrtp_frame] Allocate ZRTP frame, packet size 105
[RTPLIB][ERROR][::log_platform_error] ZRTP message length is not divisible by 32-bit word. Type: Hello   : The operation completed successfully.
 0

[RTPLIB][DEBUG][::set_zrtp_start_base] Constructed ZRTP header. Length: 22
[RTPLIB][DEBUG][::alloc_zrtp_frame] Allocate ZRTP frame, packet size 28
[RTPLIB][DEBUG][::set_zrtp_start_base] Constructed ZRTP header. Length: 3
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::alloc_zrtp_frame] Allocate ZRTP frame, packet size 105
[RTPLIB][ERROR][::log_platform_error] ZRTP message length is not divisible by 32-bit word. Type: Hello   : The operation completed successfully.
 0

[RTPLIB][DEBUG][::set_zrtp_start_base] Constructed ZRTP header. Length: 22
[RTPLIB][DEBUG][::alloc_zrtp_frame] Allocate ZRTP frame, packet size 28
[RTPLIB][DEBUG][::set_zrtp_start_base] Constructed ZRTP header. Length: 3
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::~zrtp_message] Freeing zrtp message...
[RTPLIB][DEBUG][::~zrtp_message] Freeing zrtp message...
[RTPLIB][ERROR][::init_dhm] Session initialization failed, ZRTP cannot be used!
[RTPLIB][WARNING][::init] Failed to initialize ZRTP for media stream!
[RTPLIB][ERROR][::create_stream] Failed to initialize media stream 127.0.0.1:7778/8890
[RTPLIB][DEBUG][::~zrtp_message] Freeing zrtp message...
[RTPLIB][DEBUG][::~zrtp_message] Freeing zrtp message...
[RTPLIB][ERROR][::init_dhm] Session initialization failed, ZRTP cannot be used!
[RTPLIB][WARNING][::init] Failed to initialize ZRTP for media stream!
[RTPLIB][ERROR][::create_stream] Failed to initialize media stream 127.0.0.1:8888/7776
[RTPLIB][DEBUG][::alloc_zrtp_frame] Allocate ZRTP frame, packet size 105
[RTPLIB][ERROR][::log_platform_error] ZRTP message length is not divisible by 32-bit word. Type: Hello   : The operation completed successfully.
 0

[RTPLIB][DEBUG][::set_zrtp_start_base] Constructed ZRTP header. Length: 22
[RTPLIB][DEBUG][::alloc_zrtp_frame] Allocate ZRTP frame, packet size 28
[RTPLIB][DEBUG][::set_zrtp_start_base] Constructed ZRTP header. Length: 3
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::alloc_zrtp_frame] Allocate ZRTP frame, packet size 105
[RTPLIB][ERROR][::log_platform_error] ZRTP message length is not divisible by 32-bit word. Type: Hello   : The operation completed successfully.
 0

[RTPLIB][DEBUG][::set_zrtp_start_base] Constructed ZRTP header. Length: 22
[RTPLIB][DEBUG][::alloc_zrtp_frame] Allocate ZRTP frame, packet size 28
[RTPLIB][DEBUG][::set_zrtp_start_base] Constructed ZRTP header. Length: 3
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::recv_msg] Receiving a ZRTP message
[RTPLIB][DEBUG][::~zrtp_message] Freeing zrtp message...
[RTPLIB][DEBUG][::~zrtp_message] Freeing zrtp message...
[RTPLIB][ERROR][::init_dhm] Session initialization failed, ZRTP cannot be used!
[RTPLIB][WARNING][::init] Failed to initialize ZRTP for media stream!
[RTPLIB][ERROR][::create_stream] Failed to initialize media stream 127.0.0.1:7776/8888
[RTPLIB][DEBUG][::~zrtp_message] Freeing zrtp message...
[RTPLIB][DEBUG][::~zrtp_message] Freeing zrtp message...
[RTPLIB][ERROR][::init_dhm] Session initialization failed, ZRTP cannot be used!
[RTPLIB][WARNING][::init] Failed to initialize ZRTP for media stream!
[RTPLIB][ERROR][::create_stream] Failed to initialize media stream 127.0.0.1:8890/7778
[RTPLIB][DEBUG][::~receiver] destroy receiver
[RTPLIB][DEBUG][::~receiver] destroy receiver
```

**Comment by polusto** [comment]
I also think I encountered this problem when doing my example improvements and testing the new examples.

**Comment by jrsnen** [comment]
I'll see if I can get this one fixed. My prime suspect is what happens if two sessions are started at the same time. Does this cause some sort of race condition?

**Comment by jrsnen** [comment]
I managed to figure out the reason for this bug. uvgRTP currently has race-condition when it comes to multithreaded creation of multiple media_streams with ZRTP. Which media_stream performs the Diffie-Hellman is not deterministic and this can lead to mismatches between ends as they may select different media_streams for the task which then never get completed. Concretely, the issue is which thread can first lock [this](https://github.com/ultravideo/uvgRTP/blob/177a0b0d838d3cf3d14213a1fe0bf73e67f0f1f7/src/zrtp.cc#L622) mutex.

This bug does not show up when:
- Only one media_stream is needed
- All needed media_streams are created in the same thread

Tasks:
- Fix ZRTP_Multistream example
- Document the safe usage of zrtp
- Maybe add an enum value to API that makes it possible to select the media_stream that performs DH. This enum should have the default value, which would correspond to current behavior.

**Comment by jrsnen** [comment]
Fixed by 8b05132deb206bcb1dbcbcb2861efbb417aa0b2c

I made it so that the Diffie-Hellman exchange is performed always unless the user specifies which stream does not perform DH.

ZRTP did fail once for me even after this fix, but I have been unable to replicate it. If the problem appears again, a new issue can be created. All in all, now multistream usage of ZRTP should be much more reliable.
